### PR TITLE
Image to Image Retrieval with example

### PR DESCRIPTION
**Pull request by hatianzhang**
# Description
Image to Image Retrieval using image embedding

GPT4V for understanding the correlations between retrieved images and input image

Fixes # (issue)

## Type of Change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [ ] Added new unit/integration tests
- [x] Added new notebook (that tests end-to-end)
- [x] I stared at the code and made sure it makes sense

# Suggested Checklist:

- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added Google Colab support for the newly added notebooks.
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] I ran `make format; make lint` to appease the lint gods

### CODE REVIEW

File: llama_index/core/base_multi_modal_retriever.py [hunk 1]
```diff
@@ -13,15 +13,27 @@ class MultiModalRetriever(BaseRetriever, BaseImageRetriever):
...
+    ) -> List[NodeWithScore]:
+        """Retrieve image nodes given text query.
+
+        Implemented by the user.
+
+        """
+
+    @abstractmethod
+    def image_to_image_retrieve(
+        self, str_or_query_bundle: QueryType
```
Comment (jerryjliu) [inline llama_index/core/base_multi_modal_retriever.py:34]: input type seems wrong

File: llama_index/embeddings/base.py [hunk 2]
```diff
@@ -132,6 +132,60 @@ async def aget_query_embedding(self, query: str) -> Embedding:
             )
         return query_embedding
 
+    def _get_image_embedding(self, img_file_path: ImageType) -> Embedding:
```
Comment (jerryjliu) [inline llama_index/embeddings/base.py:135]: why do we need to make these changes? we already have the `MultiModalEmbedding` class

**Review by jerryjliu — COMMENTED** [review]
notebook looks reasonable, i'd probably just spend some time fixing spelling mistakes + adding descriptions

File: llama_index/core/base_multi_modal_retriever.py (hunk 1 above)
Comment (hatianzhang) [reply llama_index/core/base_multi_modal_retriever.py:34]: oh it is as expected. query_bundle now stores image_path and we could return image_path by calling `embedding_image`

File: llama_index/indices/multi_modal/base.py [hunk 3]
```diff
@@ -85,7 +85,7 @@ def image_vector_store(self) -> VectorStore:
         return self._image_vector_store
 
     @property
-    def image_embed_model(self) -> EmbedType:
+    def image_embed_model(self) -> MultiModalEmbedding:
```
Comment (hatianzhang) [inline llama_index/indices/multi_modal/base.py:88]: @nerdai basically I changed this to MM embedding model.
As discussed offline, we will have a dedicated image_to_text_embed_model separately

**Review by jerryjliu — APPROVED** [review]

File: llama_index/indices/multi_modal/base.py (hunk 3 above)
Comment (nerdai) [reply llama_index/indices/multi_modal/base.py:88]: Got it, thanks @hatianzhang! Will check how it affects my MM RAG Eval nb and make a PR for any necessary changes.

**Pull request merged**